site-src/geps/gep-1016.md

```diff
@@ -0,0 +1,813 @@
+would require special configuration to work properly.
+
+#### Content-Based Routing
+
+While not included in the scope of this initial GEP, a common use case cited for
+routing gRPC is payload-aware routing. That is, routing rules which determine a
+backend based on the contents of the protocol buffer payload.
+
+#### User Experience
+
+The user experience would also degrade significantly if forced to route at the level of HTTP.
+
+- Encoding services and methods as URIs (an implementation detail of gRPC)
+- The [Transfer Encoding header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding) for trailers
+- Many features supported by HTTP/2 but not by gRPC, such as
+  - Query parameters
+  - Methods besides `POST`
+  - CORS
+
+
+#### Proxyless Service Mesh
+
+The gRPC library supports proxyless service mesh, a system by which routing
+configuration is received not by an in-line proxy or sidecar proxy but by the client
+itself. Eventually, `GRPCRoute` in the Gateway APIs should support this feature.
+However, to date, there are no HTTP client libraries capable of participating
+in a proxyless service mesh.
+
+--- 
+
+### Cross Serving
+
+A new route type (and especially one encapsulated in an already supported
+protocol) raises questions about precedence of routing rules. Today, there is
+already a complicated algorithm dictating which routing rule to apply to a
+particular HTTP request when two `HTTPRoutes` both apply to the request's
+hostname. What semantics should apply when a `GRPCRoute` and an `HTTPRoute` both apply to
+a request?
+
+We propose that the GRPCRoute's matchers be translated into the equivalent
+HTTPRoute matchers and that the normal HTTPRoute precedence rules apply.
+This translation is mostly straightforward, but one significant deviation between
+GRPCRoute and HTTPRoute poses difficulties: the gRPC method matcher.
+
+gRPC methods are encoded in the URI of an HTTP/2 message in the format
+`/SERVICE/METHOD`. The matcher allows the user to supply these independently, so
+that the user may leave either blank, to be interpreted as an unconditional
+match. Further, this field may either be an exact match or a regex match.
+
+To translate this field to an HTTPRoute path matcher, we propose the following
+scheme:
+
+- If neither service nor method is supplied and the type is `Exact`, translate to a
+  path match of `/` with type `PathPrefix`
+- If service and method are supplied and the type is `Exact`, translate to a
+  path match of `/SERVICE/METHOD` with type `Exact`
+- If service is supplied but not method and the type is `Exact`, translate to a
+  path match of `/SERVICE/` with type `PathPrefix`
+- Otherwise, translate the service and the method individually into regular
+  expressions (with an empty field interpreted as `.*`), and translate to an
+  HTTPRoute path matcher of type `RegularExpression` with value `/SERVICE_REGEX/MATCHER_REGEX`
```

Comment:
There is one missing piece - the protocol. We don't have an explicit HTTP/2 protocol and our current guidance is loose:

https://github.com/kubernetes-sigs/gateway-api/blob/3c8ea05823d3a4215f86b9b3d56316bf2e74a066/apis/v1alpha2/gateway_types.go#L265-L272

One could translate a gRPCRoute to HTTPRoute but if the underlying gateway doesn't support HTTP/2 or has it disabled, none of this would work.

Reply:
I see. I was under the impression that the `HTTP` protocol implied `HTTP2`. I will firm this up.